[REEF-1023] Handle Exceptions in Vortex callback

afchung opened the pull request:
This addressed the issue by
  * Use ``FutureCallback`` instead of ``EventHandler``.
  * Added tests for tasklet exception scenarios.

JIRA:
  [REEF-1023](https://issues.apache.org/jira/browse/REEF-1023)

yunseong commented on lang/java/reef-applications/reef-vortex/src/test/java/org/apache/reef/vortex/driver/DefaultVortexMasterTest.java:
```diff
@@ -156,6 +162,49 @@ public void testMultipleTaskletsFailure() throws Exception {
+    assertTrue("Callback should have been received", callbackReceived.get());
+  }
+
+
```
Nit: this additional line can be removed.

yunseong commented on lang/java/reef-applications/reef-vortex/src/test/java/org/apache/reef/vortex/driver/DefaultVortexMasterTest.java:
```diff
@@ -156,6 +162,49 @@ public void testMultipleTaskletsFailure() throws Exception {
     }
   }
 
+
```
Nit: this additional line can be removed.

yunseong commented:
@afchung I've looked through this change and it looks good. @bgchun Could you take another look?

bgchun commented:
@afchung This looks good. I'd like to suggest one thing, though. Instead of taking the guava dependency, it'd be better to create our own `FutureCallback` in `o.a.r.vortex.api` without introducing the dependency.

afchung commented:
@yunseong @bgchun Thanks for the review! I've addressed the comments and also switched to use threads for the callbacks for asynchronicity, please have another look. Thanks!

yunseong commented:
@afchung Having our own callback looks great! I have one question: would it be better to use `Executors` than creating `new Thread()`? I'm not sure the benefit though, because we create only one thread per callback. @bgchun What do you think?

bgchun commented:
@yunseong Good point. Each VortexFuture callback creates a thread, which seems to be heavy weight if we launch many tasklets. It'd be better to create a shared executor for handling the callbacks. The executor is shared among `VortexFuture`s. 
@afchung How does this sounds?

afchung commented:
Sounds good. My original thought was that creating a thread for each tasklet doesn't seem to be too much, but perhaps using an Executor will be more efficient.

johnyangk commented:
+1 for fixed number of threads, as many tasklets are expected.

afchung commented:
@yunseong @bgchun @johnyangk Updated to use ``Executor``. Thanks!

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexMasterConf.java:
```diff
@@ -67,6 +67,13 @@
+   * Size of threadpool for callbacks on {@link org.apache.reef.vortex.api.VortexFuture}.
+   */
+  @NamedParameter(doc = "Size of threadpool for callbacks on VortexFuture.", default_value = "10")
+  final class CallbackThreadpoolSize implements Name<Integer> {
```
Nit: We use capital P elsewhere (e.g., `ThreadPool`), so it would be better to be `CallbackThreadPoolSize`.

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexMasterConf.java:
```diff
@@ -97,6 +104,11 @@
   /**
+   * Size of threadpool for callbacks on VortexFuture.
+   */
+  public static final OptionalParameter<Integer> FUTURE_CALLBACK_THREADPOOL_SIZE = new OptionalParameter<>();
```
Here also, as we use `ThreadPool`, I think this would be `FUTURE_CALLBACK_THREAD_POOL_SIZE`

yunseong commented:
Thanks @afchung. The change looks great to me. I left some minor comments (I left the comments at your forked branch, and wrote again. Sorry for the inconvenience).

bgchun commented:
@afchung This looks good. I agree with @yunseong on naming: `ThreadPool`. After your addressing it, I will test and merge.

afchung commented:
@yunseong @bgchun That's fair; I've made the changes, please have a look. Thanks!
